advanced/sr_gui_fingertip_visualization/src/sr_gui_fingertip_visualization/visual_tab_layouts.py

```diff
@@ -0,0 +1,439 @@
+            layout.addWidget(self._finger_widgets[finger], alignment=Qt.AlignCenter)
+
+        elif detected_fingertip_type == BiotacType.BLANK:
+            self._finger_widgets[finger] = QLabel("No tactile sensor")
+            layout.addWidget(self._finger_widgets[finger], alignment=Qt.AlignCenter)
+
+        self._finger_frame[finger].setLayout(layout)
+        return self._finger_frame[finger]
+
+    def _detect_biotac_type(self, finger):
+        sum_of_electrode_values = sum(self._data[finger]['electrodes'])
+        biotac_type = None
+        if sum_of_electrode_values == 0:
+            biotac_type = BiotacType.BLANK
+        elif sum_of_electrode_values == self._electrode_count:
+            biotac_type = BiotacType.SP_MINUS
+        else:
+            biotac_type = BiotacType.SP_PLUS
+        return biotac_type
+
+    def start_timer_and_subscriber(self):
+        self._subscriber = rospy.Subscriber('/{}/tactile'.format(self._side), BiotacAll, self._tactile_data_callback)
+        self._timer.timeout.connect(self.timerEvent)
+        self._timer.start(10)
+
+    def stop_timer_and_subscriber(self):
+        self._timer.stop()
+        if self._subscriber:
+            self._subscriber.unregister()
+
+    def timerEvent(self):
+        for finger in self._fingers:
+            if self._finger_frame[finger].isChecked():
+                if type(self._finger_widgets[finger]) == list:
+                    for electrode in range(self._electrode_count):
+                        if type(self._finger_widgets[finger][electrode]) == DotUnitBiotacSPPlus:
+                            if electrode < len(self._data[finger][self._datatype_to_display]):
+                                try:
+                                    value = self._data[finger][self._datatype_to_display][electrode]
+                                    self._finger_widgets[finger][electrode].update_data(value)
+                                    self._finger_widgets[finger][electrode].show()
+                                except Exception:
+                                    pass
+                            else:
+                                self._finger_widgets[finger][electrode].hide()
+
+                    self._data_labels[finger]['pac0'].setText("pac0:{}".format(self._data[finger]['pac0']))
+                    self._data_labels[finger]['pac1'].setText("pac1:{}".format(self._data[finger]['pac1']))
+                    self._data_labels[finger]['pdc'].setText("pdc:{}".format(self._data[finger]['pdc']))
+                    self._data_labels[finger]['tac'].setText("tac:{}".format(self._data[finger]['tac']))
+                    self._data_labels[finger]['tdc'].setText("tdc:{}".format(self._data[finger]['tdc']))
+
+                elif type(self._finger_widgets[finger]) == DotUnitBiotacSPMinus:
+                    self._finger_widgets[finger].update_data(self._data[finger])
+
+
+class VisualizationTab(QWidget):
```

Comment:
Once again I see a lot of duplicated code ... Might be wort having a bit more generic class